Make blockstorage globals private members of BlockManager

maflcko opened the pull request:
Globals aren't too nice because they hide dependencies, also they make testing harder.

Fix that by removing some.

Commit:
Move functions to BlockManager

Needed for a later commit

Commit:
Make blockstorage globals private members of BlockManager

Commit:
move-only: Create WriteBlockIndexDB helper

Can be reviewed with --color-moved=dimmed-zebra --color-moved-ws=ignore-all-space

ryanofsky commented on src/node/blockstorage.cpp:
```diff
@@ -348,6 +331,11 @@ void BlockManager::Unload()
 
     m_block_index.clear();
+
+    vinfoBlockFile.clear();
+    nLastBlockFile = 0;
+    setDirtyBlockIndex.clear();
+    setDirtyFileInfo.clear();
```
In commit "Make blockstorage globals private members of BlockManager" (fad81b7a0f8ea300708321c3fcae7f75cade14ba)

Is this a pure refactor, or is there some change in behavior here? It seems like these lines are being moved from `UnloadBlockIndex` to `BlockManager::Unload` but neither function is directly calling the other.

It might be good to move this change to a separate commit before the "make private members" commit with an explanation about why it's safe. Not saying it's unsafe. It's just not obviously safe.

MarcoFalke replied:
`UnloadBlockIndex` calls `chainman.Unload()`, which again should call `BlockManager::Unload`. So only change should be the destruction order

ryanofsky replied:
> `UnloadBlockIndex` calls `chainman.Unload()`, which again should call `BlockManager::Unload`. So only change should be the destruction order

So there is no code in some other place calling `BlockManager::Unload` and now having extra state cleared?

If you put this change in a separate commit and explicitly say whether it's _supposed_ to change behavior and maybe say how it doesn't change behavior, it can make it easier for future reviewers and future regression bugfixers, who I think would logically be asking these questions.

MarcoFalke replied:
Thanks Ryan. Reworked the pull into more commits with descriptions in the body.

ryanofsky replied:
> Thanks Ryan.

Ok, you finally convinced me

Commit:
scripted-diff: Rename touched member variables

-BEGIN VERIFY SCRIPT-

 ren() { sed -i "s/\<$1\>/$2/g" $( git grep -l "$1" ./src/ ) ; }

 ren vinfoBlockFile     m_blockfile_info
 ren nLastBlockFile     m_last_blockfile
 ren fCheckForPruning   m_check_for_pruning
 ren setDirtyBlockIndex m_dirty_blockindex
 ren setDirtyFileInfo   m_dirty_fileinfo

-END VERIFY SCRIPT-

Commit:
Move blockstorage-related unload to BlockManager::Unload

This is a refactor and safe to do because:
* UnloadBlockIndex calls ChainstateManager::Unload, which calls
  BlockManager::Unload
* Only unit tests call Unload directly

Commit:
test: Load genesis block to allow flush

This is needed to turn globals into member variables. Otherwise, this
will lead to issues:

runtime error: reference binding to null pointer of type 'CBlockFileInfo'
    #0 in std::vector<CBlockFileInfo, std::allocator<CBlockFileInfo> >::operator[](unsigned long) /usr/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/stl_vector.h:1046:2
    #1 in BlockManager::FlushBlockFile(bool, bool) src/node/blockstorage.cpp:540:47
    #2 in CChainState::FlushStateToDisk(BlockValidationState&, FlushStateMode, int) src/validation.cpp:2262:28
    #3 in CChainState::ResizeCoinsCaches(unsigned long, unsigned long) src/validation.cpp:4414:15
    #4 in validation_chainstate_tests::validation_chainstate_resize_caches::test_method() src/test/validation_chainstate_tests.cpp:66:12

Commit:
doc: Clarify nPruneAfterHeight for signet

ryanofsky reviewed (approved):
Code review ACK fa68a6c2fc6754c160e0f98007785602201b3c47. Nice changes!

Sjors commented:
ACK facd3df21f344dd84e5f28862056700c1fded17c

jonatack commented on src/node/blockstorage.cpp:
```diff
@@ -348,6 +331,31 @@ void BlockManager::Unload()
+}
+
+bool BlockManager::WriteBlockIndexDB()
+{
```
fa467f3913918701 Missing `AssertLockHeld(::cs_main)` in the definition of this new helper. Added in #24002.

jonatack commented on src/node/blockstorage.h:
```diff
@@ -120,6 +142,16 @@ class BlockManager
 
     CBlockIndex* LookupBlockIndex(const uint256& hash) const EXCLUSIVE_LOCKS_REQUIRED(cs_main);
 
+    /** Get block file info entry for one block file */
```
Would be it worth mentioning that this function is only used by the unit tests? If yes, can add to #24002. 

```diff
-    /** Get block file info entry for one block file */
+    /** Get block file info entry for one block file. Currently only used for unit tests. */
```

MarcoFalke replied:
It might be better to move the function to the test, if it is only used there.
